codegen/src/SchemaProvider.scala

```diff
@@ -2,89 +2,144 @@ package besom.codegen
-  def packageInfo(providerName: ProviderName, schemaVersion: SchemaVersion): PulumiPackageInfo
 
-  def pulumiPackage(schemaFilePath: os.Path): PulumiPackage = {
-    PulumiPackage.fromFile(schemaFilePath)
-  }
-
-  def loadPackageInfo(pulumiPackage: PulumiPackage): PulumiPackageInfo = {
-    val enumTypeTokensBuffer   = ListBuffer.empty[String]
-    val objectTypeTokensBuffer = ListBuffer.empty[String]
+  def dependencies(schemaName: SchemaName, packageVersion: PackageVersion): List[(SchemaName, PackageVersion)]
 
-    // post-process the package to improve its quality
-    pulumiPackage.types.foreach {
-      case (typeToken, _: EnumTypeDefinition) =>
-        enumTypeTokensBuffer += typeToken.toLowerCase // Unifying to lower case to circumvent inconsistencies in low quality schemas (e.g. aws)
-      case (typeToken, _: ObjectTypeDefinition) =>
-        objectTypeTokensBuffer += typeToken.toLowerCase
-    }
-
-    PulumiPackageInfo(
-      enumTypeTokens = enumTypeTokensBuffer.toSet,
-      objectTypeTokens = objectTypeTokensBuffer.toSet,
-      resourceTypeTokens = pulumiPackage.resources.keySet.map(_.toLowerCase),
-      moduleToPackageParts = pulumiPackage.moduleToPackageParts,
-      providerToPackageParts = pulumiPackage.providerToPackageParts
-    )
-  }
+  def packageInfo(schemaOrMetadata: Either[SchemaFile, PackageMetadata]): (PulumiPackage, PulumiPackageInfo)
+  def packageInfo(metadata: PackageMetadata): (PulumiPackage, PulumiPackageInfo)
 }
 
 class DownloadingSchemaProvider(schemaCacheDirPath: os.Path)(implicit logger: Logger) extends SchemaProvider {
 
-  import SchemaProvider._
-
-  private val packageInfos: collection.mutable.Map[(ProviderName, SchemaVersion), PulumiPackageInfo] =
+  private val schemaFileName = "schema.json"
+  private val packageInfos: collection.mutable.Map[(SchemaName, PackageVersion), PulumiPackageInfo] =
     collection.mutable.Map.empty
 
-  private def downloadedSchemaFilePath(providerName: ProviderName, schemaVersion: SchemaVersion): os.Path = {
-    val schemaFilePath = schemaCacheDirPath / providerName / schemaVersion / "schema.json"
+  private def pulumiPackage(metadata: PackageMetadata): (PulumiPackage, PackageMetadata) = {
+    val schemaFilePath = schemaCacheDirPath / metadata.name / metadata.version / schemaFileName
 
     if (!os.exists(schemaFilePath)) {
-      logger.debug(s"Downloading schema for $providerName:$schemaVersion' into '${schemaFilePath.relativeTo(os.pwd)}'")
-      val schemaSource = s"$providerName@$schemaVersion"
-      os.makeDir.all(schemaFilePath / os.up)
-      os.proc("pulumi", "package", "get-schema", schemaSource).call(stdout = schemaFilePath)
+      logger.debug(
+        s"Downloading schema for '${metadata.name}:${metadata.version}' into '${schemaFilePath.relativeTo(os.pwd)}'"
+      )
+
+      val schemaSource =
+        if (metadata.version.isDefault)
+          metadata.name
```

Comment:
I have already experienced some problems caused by not specifying the exact version for a schema so I think we should avoid that

Reply:
what kind of problems? some schemas does not have version inside

Reply:
I'd rather try to resolve the latest version, I've corrected the logic to behave that way